### DISCUSSION TITLE
Moved from win10/python 3.10 to win11/python 3.12.3 and get: The current thread apartment type is not MTA: MAIN_STA...

### DISCUSSION THREAD

**Discussion by vascolp** [hbldh/bleak#1564] [Q&A] [answered]
Hello everybody,

I have a program that uses kivy and bleak, so bleak is ran in a separate asyncio task.
No threads. 
I was on windows 10 and python 3.10.something. 
Everything was fine.
I had to upgrade to windows 11 and installed python 3.12.3, bleak 0.22.0.
Now, when I call bleak in a separate task, I get:
"BleakError: The current thread apartment type is not MTA: MAIN_STA. Beware of packages like pywin32 that may change the apartment type implicitly."
But in another program with no tasks, calling bleak directly works just fine.
I am aware of #1556, but that seems to be just the exception throwing.
What is my problem? How can I force the program to use MTA?  
Do I have to go back to python 3.10?

Thanks for any help!!
VascoLP

**Comment by dlech** [comment]
https://bleak.readthedocs.io/en/latest/troubleshooting.html#not-working-when-threading-model-is-sta

does this help?

**Reply by dlech** [reply]
Although, using the workaround from the troubleshooting docs might break Kivy if Kivy needs the main thread to be STA for the UI.

So we might need to add an option to avoid the exception in cases like this.

**Reply by dlech** [reply]
I opened https://github.com/hbldh/bleak/issues/1565 to track this.

**Reply by dlech** [reply]
As a workaround, you can use an older version of Bleak.

**Reply by vascolp** [reply]
That solution of the documentation does not work. I tried to place it in several places... no luck.
Also, you have a typo in the documentation. It is not **utils**, it is **util**.
I also tried with python 3.10, no luck either. 
Will try with older bleak... 0.20.2.

**Reply by vascolp** [reply]
```
python -m pip install bleak==0.20.2:
...
    ERROR: Failed building wheel for bleak-winrt
    Failed to build bleak-winrt
    ERROR: Could not build wheels for bleak-winrt, which is required to install pyproject.toml-based projects

```

```
python -m pip install bleak==0.21:
    BleakError: The current thread apartment type is not MTA: MAIN_STA. 

```
Am I stuck?

**Reply by dlech** [reply]
Not sure why it was trying to build bleak-winrt instead of downloading a binary wheel for 0.20.2.

You shouldn't be seeing that error in Bleak 0.21.x. I think it was still using cached version of 0.22.0 in that case.

I just opened https://github.com/hbldh/bleak/pull/1567 with a proposed solution.

**Reply by dlech** [reply, marked as answer]
To try it, I think you can `pip install https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip` and then paste the following code snippet in your main python file.

```python
try:
    from bleak.backends.winrt.util import allow_sta
    # tell Bleak we are using a graphical user interface that has been properly
    # configured to work with asyncio
    allow_sta()
except ImportError:
    # other OSes and older versions of Bleak will raise ImportError which we
    # can safely ignore
    pass
```

**Reply by vascolp** [reply]
Yes! It works!
Thanks!!!

**Reply by dlech** [reply]
Great, thanks for testing!

**Reply by shortcakesweets** [reply]
Hi, I'm having simillar problems. Can you provide the above pip install link again? The current one (https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip) gives 404 error.

**Reply by dlech** [reply]
This has gone through a few iterations since then. Can you please start a new discussion with a minimal reproducible test case that triggers the problem you are having?

**Comment by mantisbug** [comment]
> To try it, I think you can `pip install https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip` 

I tried:
> pip install https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip

Result:
> PS C:\PyCharmProjects\companion_em_laptop> pip install https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-s
> ta.zip
> Collecting https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip
>   ERROR: HTTP error 404 while getting https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip
> ERROR: Could not install requirement https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip because of 
> HTTP error 404 Client Error: Not Found for url: https://codeload.github.com/hbldh/bleak/zip/refs/heads/winrt-allow-sta
>  for URL https://github.com/hbldh/bleak/archive/refs/heads/winrt-allow-sta.zip

I think that update has not arrived in the UK...

**Reply by vascolp** [reply]
Ok that was only a test version that was included in the latest release.
Upgrade to 0.22.1 and do the thing:

```
try:
    from bleak.backends.winrt.util import allow_sta
    # tell Bleak we are using a graphical user interface that has been properly
    # configured to work with asyncio
    allow_sta()
except ImportError:
    # other OSes and older versions of Bleak will raise ImportError which we
    # can safely ignore
    pass

```

**Reply by mantisbug** [reply]
Thanks, @vascolp, I did so. 
My main problem is that my script does not work anymore since I upgraded Bleak to 0.22.1
Working on it.

**Reply by mantisbug** [reply]
I downgraded Bleak to 0.21.1 and it works again!
